include/boost/url/impl/decoded_view.hpp

```diff
@@ -0,0 +1,323 @@
+decoded_view::const_reverse_iterator
+decoded_view::
+crbegin() const noexcept
+{
+    return rbegin();
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reverse_iterator
+decoded_view::
+rend() const noexcept
+{
+    return const_reverse_iterator(begin());
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reverse_iterator
+decoded_view::
+crend() const noexcept
+{
+    return rend();
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reference
+decoded_view::
+front() const
+{
+    BOOST_ASSERT( !empty() );
+    return *begin();
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reference
+decoded_view::
+back() const
+{
+    BOOST_ASSERT( !empty() );
+    return *rbegin();
+}
+
+BOOST_CXX14_CONSTEXPR inline
+decoded_view::size_type
+decoded_view::
+copy(
+    char* dest,
+    size_type count,
+    size_type pos) const
+{
+    if( pos > size() )
+        detail::throw_invalid_argument(
+            BOOST_CURRENT_LOCATION);
+    std::size_t rlen = (std::min)(count, size() - pos);
+    auto first = begin();
+    while (pos--)
+        ++first;
+    auto last = first;
+    while (rlen--)
+        ++last;
+    for (; first != last; ++first, ++dest)
+        *dest = *first;
+    return rlen;
+}
+
+BOOST_CXX14_CONSTEXPR inline
+int
+decoded_view::
+compare(string_view other) const noexcept
+{
+    std::size_t rlen = (std::min)(size(), other.size());
+
+    // int cmp = traits_type::compare(encoded_data(), other.data(), rlen);
+    auto first1 = begin();
+    auto last1 = end();
+    auto first2 = other.begin();
+    auto last2 = other.end();
+    int cmp = 0;
+    for ( ;
+         (first1 != last1) && (first2 != last2);
+         ++first1, ++first2, --rlen)
+    {
+        if (*first1 < *first2)
+        {
+            cmp = -1;
+            break;
+        }
+        else if (*first2 < *first1)
+        {
+            cmp = +1;
+            break;
+        }
+    }
+    if( cmp != 0 )
+        return cmp;
+    if( size() == other.size() )
+        return 0;
+    return size() < other.size()? -1: +1;
+}
+
+BOOST_CXX14_CONSTEXPR inline
+int
+decoded_view::
+compare(decoded_view other) const
+{
+    std::size_t rlen = (std::min)(size(), other.size());
+
+    // int cmp = traits_type::compare(encoded_data(), other.data(), rlen);
+    auto first1 = begin();
+    auto last1 = end();
+    auto first2 = other.begin();
+    auto last2 = other.end();
+    int cmp = 0;
+    for ( ;
+         (first1 != last1) && (first2 != last2);
+         ++first1, ++first2, --rlen)
+    {
+        if (*first1 < *first2)
+        {
+            cmp = -1;
+            break;
+        }
+        else if (*first2 < *first1)
+        {
+            cmp = +1;
+            break;
+        }
+    }
+    if( cmp != 0 )
+        return cmp;
+    if( size() == other.size() )
+        return 0;
+    return size() < other.size()? -1: +1;
+}
+
+template <class T>
+T&
+decoded_view::
+append_to(
+    T& str) const
+{
+    str.append(begin(), end());
```

Comment:
At some point a `BOOST_STATIC_ASSERT` would be friendly here, and the javadoc needs to explain what the requirements for `T` are.